### ISSUE TITLE
 Identify general code in ramp fitting

### ISSUE THREAD

**Issue by stscijgbot-rstdms** [spacetelescope/romancal#190]
_Issue [RCAL-138](https://jira.stsci.edu/browse/RCAL-138) was created on JIRA by [David Grumm](https://jira.stsci.edu/secure/ViewProfile.jspa?name=grumm):_

Review the current JWST ramp fitting step code to identify areas that can be moved to a common area and be used by both the JWST and Roman pipelines.

**Comment by stscijgbot-rstdms** [comment]
_Comment by [David Grumm](https://jira.stsci.edu/secure/ViewProfile.jspa?name=grumm) on [JIRA](https://jira.stsci.edu/browse/RCAL-138?focusedCommentId=517967#comment-517967):_

With some minor API changes, the attached file is the same version of the top-level driver ramp_fit.step.py that I committed for RCAL-72 (PR#&#x2060;83).  This version is the same as used for JWST version, except for the additions in the areas noted. The 2 "Roman only" blocks delineate code that does not appear in the JWST code. [^roman_ramp_fit_step.py]

**Comment by stscijgbot-rstdms** [comment]
_Comment by [David Grumm](https://jira.stsci.edu/secure/ViewProfile.jspa?name=grumm) on [JIRA](https://jira.stsci.edu/browse/RCAL-138?focusedCommentId=518070#comment-518070):_

To get an idea of the generality of the ramp fitting step code, I reviewed the functions in the 2 modules (ramp_fits.py & utils.py) containing functions used by the step. For the 30 functions in ramp_fit.py, these are the percentages of the functions in 4 categories:                                                                                        
                                                                                                                       
1) The function is general and could be moved to a common area for use by both missions: 37%                           
                                                                                                                       
2) The function uses JWST datamodels (some of which may not be compatible with their Roman counterparts): 26%          
                                                                                                                       
3) The function has other JWST-specific (usually MIRI) code: 13%                                                       
                                                                                                                       
4) The function uses other, non-datamodel code frequently depending on mission-specific arrays, with specific shapes or dimensions : 24%                                                                                                     
                                                                                                                       
In the above categories, there is some overlap which I have ignored for simplicity.                                    
                                                                                                                       
On a separate topic, this file has ~3200 lines, and could benefit from refactoring.                                                           
I will next do a similar review of utils.py.

**Comment by stscijgbot-rstdms** [comment]
_Comment by [David Grumm](https://jira.stsci.edu/secure/ViewProfile.jspa?name=grumm) on [JIRA](https://jira.stsci.edu/browse/RCAL-138?focusedCommentId=518480#comment-518480):_

For the 26 functions in utils.py, these are the percentages of the functions in 4 categories:                                                       
                                                                                                                                                    
1) The function is general and could be moved to a common area for use by both missions: 11%                                                        
                                                                                                                                                    
2) The function uses JWST datamodels (some of which may not be compatible with their Roman counterparts): 65%                                       
                                                                                                                                                    
3) The function has other JWST-specific (usually MIRI) code: 0%                                                                                     
                                                                                                                                                    
4) The function uses other, non-datamodel code frequently depending on mission-specific arrays, with specific shapes or dimensions: 23%             
                                                                                                                                                    
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -                                     
                                                                                                                                                    
For all 56 functions in the modules ramp_fit.py & utils.py, these are the percentages of the functions in 4 categories:                             
                                                                                                                                                    
1) The function is general and could be moved to a common area for use by both missions: 25%                                                        
                                                                                                                                                    
2) The function uses JWST datamodels (some of which may not be compatible with their Roman counterparts): 45%                                       
                                                                                                                                                    
3) The function has other JWST-specific (usually MIRI) code: 7%                                                                                     
                                                                                                                                                    
4) The function uses other, non-datamodel code frequently depending on mission-specific arrays, with specific shapes or dimensions: 23%             
                                                                                                                                                    
In summary, about 1/4 of the ramp fitting functions are sufficiently general to be moved to a common area without modification, about 1/2 use JWST DataModels so would need some work to enable them to handle multimission data, and the remaining 1/4 use numpy arrays having a dimension for the integration number, which could be handled by resizing the relevant Roman data arrays.
